### PR TITLE
Allow transactions to be discarded

### LINKED ISSUE #120: Add way to skip requests by external monitoring tools
Tools (such as Pingdom) can generate quite a few requests that add up to the plan limit, but don't offer any insights. 

Some kind of way to tell that a transaction can be discarded by the gem would be nice. 

Requested through Intercom: https://app.intercom.io/a/apps/yzor8gyw/inbox/all/conversations/5481377438

### PR DESCRIPTION
**Pull request by jeroenvisser101**
Fixes #120

### CODE REVIEW

File: spec/lib/appsignal/transaction_spec.rb [hunk 1]
```diff
@@ -120,6 +120,17 @@ def []=(key, val)
...
       end
+
+      context "if a transaction is discarded" do
+        it "should not complete the transaction" do
+          original_transaction = Appsignal::Transaction.current
+          Appsignal::Transaction.current.should_not_receive(:complete)
+
+          Appsignal::Transaction.complete_current!
+
+          Thread.current[:appsignal_transaction].should == original_transaction
```
Comment (jeroenvisser101) [inline spec/lib/appsignal/transaction_spec.rb]: Should this also set the appsignal_transaction to nil?

**Comment by jeroenvisser101** [comment]
I think it's stuck, retriggering the tests

**Comment by matsimitsu** [comment]
Looks good! Maybe add a spec that calls these methods:

```ruby
def discard!
  @discarded = true
end

def restore!
  @discarded = false
end

def discarded?
  @discarded == true
end
```
Just to prevent breakage in the future :)

**Comment by jeroenvisser101** [comment]
Done and green! :)

**Comment by matsimitsu** [comment]
👍 @tombruijn what do you think :)

**Comment by tombruijn** [comment]
Looks good!

I think it will need another PR if we want to automatically discard transactions for certain requests as specified in #120. Or will we document a way to use this in the docs?

**Comment by matsimitsu** [comment]
I'd leave that up to the user (in a before/after controller action), so documenting a way to do this in the docs is a 👍  for me.

**Comment by jeroenvisser101** [comment]
@tombruijn thanks! Yeah, I would not want Appsignal to decide that, as we might want to have that on sometimes. And we would like to define our own cases as where to discard a request.

**Comment by tombruijn** [comment]
Alright, that makes sense :)

So a (documentation) use case would be

```ruby
class PostController < ApplicationController
  before_action :ignore_bad_bots, :allow_safe_bots

  def index
    # ...
  end

  private

  def ignore_bad_bots
    if request.user_agent =~ /bot/
      Appsignal::Transaction.current.discard!
    end
  end

  def allow_safe_bots
    if request.user_agent =~ /allowed bot/
      Appsignal::Transaction.current.restore!
    end
  end
end
```

Since an average gem user won't really know what a Transaction is, and I don't think we want to have them interface with that class, can we make a little convenience method for this? Like we have `Appsignal.without_instrumentation` for `Appsignal::Transaction#pause` and `Appsignal::Transaction#resume`, we can have a `Appsignal.ignore_request|action!` for example

**Comment by jeroenvisser101** [comment]
@tombruijn we could do that, although this might not be something that an average user would implement.

**Comment by jeroenvisser101** [comment]
and documentation is good indeed. The restore could also be called in a `rescue_from` call to make sure transactions are sent when errors occur.

**Comment by tombruijn** [comment]
Good idea on the rescue_from. Updated my comment's example.

@matsimitsu what do you think on the convenience methods for the discard/restore! methods? I think it's better not to let users directly interface with the Transaction class.

I'm okay to merge this 👍  and add convenience methods at a later time

**Comment by thijsc** [comment]
This looks good, we could add convenience methods later.

**Comment by jeroenvisser101** [comment]
Thanks @thijsc!

**Comment by thijsc** [comment]
This has now been released in 1.3.2.

**Pull request merged**